### ISSUE TITLE
ACE and OSX

### ISSUE THREAD

**Issue by rijobro** [CCPPETMR/SIRF-SuperBuild#186]
I'm installing the superbuild on OSX. 

I have ACE installed via brew (`brew install ace`), but when I enter `USE_SYSTEM_ACE=ON` and `ACE_DIR=/usr/local/Cellar/ace/6.5.3/`, upon configuring `ACE_DIR` is reset to `ACE_DIR-NOTFOUND`.

So I tried with `USE_SYSTEM_ACE=OFF`, but I came across the following error:

```
[ 23%] Performing update step for 'ACE'
HEAD is up to date.
[ 24%] Performing configure step for 'ACE'
loading initial cache file /Users/rich/Documents/Code/SIRF-Superbuild/Build/builds/ACE/tmp/ACE-cache-.cmake
-- Configuring done
-- Generating done
-- Build files have been written to: /Users/rich/Documents/Code/SIRF-Superbuild/Build/builds/ACE/build
[ 25%] Performing build step for 'ACE'
[ 12%] Performing build step for 'ACE'
linux
6.4.7
INSTALL_DIR  /Users/rich/Documents/Code/SIRF-Superbuild/Build/INSTALL/lib
INCLUDE_DIR  /Users/rich/Documents/Code/SIRF-Superbuild/Build/INSTALL/include
/Users/rich/Documents/Code/SIRF-Superbuild/Build/builds/ACE/build/source

GNUmakefile: /Users/rich/Documents/Code/SIRF-Superbuild/Build/builds/ACE/build/source/ace/GNUmakefile.ACE MAKEFLAGS=s --jobserver-fds=4,5 -j

clang: error: no such file or directory: 'Unknown'
make[7]: *** [.shobj/ACE.o] Error 1
make[6]: *** [ACE] Error 2
make[6]: *** Waiting for unfinished jobs....
ACE_Qt4Reactor_moc will not be built due to the following disabled make macros: qt4

GNUmakefile: /Users/rich/Documents/Code/SIRF-Superbuild/Build/builds/ACE/build/source/ace/QtReactor/GNUmakefile.ACE_Qt4Reactor_moc MAKEFLAGS=s --jobserver-fds=4,5 -j
```

Any ideas?

**Comment by paskino** [comment]
It's difficult to debug ACE's build. It seems there is a target called `Unknown`. Or an env variable.

Can you please try to `make ACE VERBOSE=2`?

**Comment by paskino** [comment]
But, do you need to set the `ACE_DIR` variable?

**Comment by rijobro** [comment]
Here it is with `verbose=2`. Looks like the problematic line is `-DACE_LACKS_LINUX_NPTL Unknown`.

```
eduroam-int-dhcp-97-123-73:build rich$ make VERBOSE=2
/Applications/CMake.app/Contents/bin/cmake -H/Users/rich/Documents/Code/SIRF-Superbuild/Build/sources/ACE -B/Users/rich/Documents/Code/SIRF-Superbuild/Build/builds/ACE/build --check-build-system CMakeFiles/Makefile.cmake 0
/Applications/CMake.app/Contents/bin/cmake -E cmake_progress_start /Users/rich/Documents/Code/SIRF-Superbuild/Build/builds/ACE/build/CMakeFiles /Users/rich/Documents/Code/SIRF-Superbuild/Build/builds/ACE/build/CMakeFiles/progress.marks
/Applications/Xcode.app/Contents/Developer/usr/bin/make -f CMakeFiles/Makefile2 all
/Applications/Xcode.app/Contents/Developer/usr/bin/make -f CMakeFiles/ACE.dir/build.make CMakeFiles/ACE.dir/depend
cd /Users/rich/Documents/Code/SIRF-Superbuild/Build/builds/ACE/build && /Applications/CMake.app/Contents/bin/cmake -E cmake_depends "Unix Makefiles" /Users/rich/Documents/Code/SIRF-Superbuild/Build/sources/ACE /Users/rich/Documents/Code/SIRF-Superbuild/Build/sources/ACE /Users/rich/Documents/Code/SIRF-Superbuild/Build/builds/ACE/build /Users/rich/Documents/Code/SIRF-Superbuild/Build/builds/ACE/build /Users/rich/Documents/Code/SIRF-Superbuild/Build/builds/ACE/build/CMakeFiles/ACE.dir/DependInfo.cmake --color=
/Applications/Xcode.app/Contents/Developer/usr/bin/make -f CMakeFiles/ACE.dir/build.make CMakeFiles/ACE.dir/build
[ 12%] Performing build step for 'ACE'
cd /Users/rich/Documents/Code/SIRF-Superbuild/Build/builds/ACE/build/build && bash /Users/rich/Documents/Code/SIRF-Superbuild/Build/builds/ACE/build/build_ace.sh -p linux -s /Users/rich/Documents/Code/SIRF-Superbuild/Build/builds/ACE/build/source -j 2 -l /Users/rich/Documents/Code/SIRF-Superbuild/Build/INSTALL/lib -i /Users/rich/Documents/Code/SIRF-Superbuild/Build/INSTALL/include
linux
6.4.7
INSTALL_DIR  /Users/rich/Documents/Code/SIRF-Superbuild/Build/INSTALL/lib
INCLUDE_DIR  /Users/rich/Documents/Code/SIRF-Superbuild/Build/INSTALL/include
/Users/rich/Documents/Code/SIRF-Superbuild/Build/builds/ACE/build/source

GNUmakefile: /Users/rich/Documents/Code/SIRF-Superbuild/Build/builds/ACE/build/source/ace/GNUmakefile.ACE MAKEFLAGS= --jobserver-fds=4,5 -j -- VERBOSE=2

c++ -fvisibility=hidden -fvisibility-inlines-hidden -Wnon-virtual-dtor -O3 -ggdb -pthread -fno-strict-aliasing -Wall -W -Wpointer-arith -pipe -DACE_LACKS_LINUX_NPTL Unknown -D_GNU_SOURCE   -I/Users/rich/Documents/Code/SIRF-Superbuild/Build/builds/ACE/build/source/ -D__ACE_INLINE__ -I.. -DACE_BUILD_DLL  -c -fPIC -o .shobj/ACE.o /Users/rich/Documents/Code/SIRF-SuperBuild/Build/builds/ACE/build/source/ace/ACE.cpp
clang: error: no such file or directory: 'Unknown'
make[4]: *** [.shobj/ACE.o] Error 1
make[3]: *** [ACE] Error 2
make[3]: *** Waiting for unfinished jobs....
ACE_Qt4Reactor_moc will not be built due to the following disabled make macros: qt4

GNUmakefile: /Users/rich/Documents/Code/SIRF-Superbuild/Build/builds/ACE/build/source/ace/QtReactor/GNUmakefile.ACE_Qt4Reactor_moc MAKEFLAGS= --jobserver-fds=4,5 -j -- VERBOSE=2
```

**Comment by rijobro** [comment]
> But, do you need to set the ACE_DIR variable?

You do if you want to use `USE_SYSTEM_ACE=ON`.

**Comment by rijobro** [comment]
I don't think it's related to Gadgetron, since `make ACE` gives the error (not `make gadgetron`).

**Comment by KrisThielemans** [comment]
There's a few different things here.

# ACE build failure

The way we build ACE seems a bit over-complicated, but then i haven't checked it properly. Essentially, SIRF-SuperBuild clones https://github.com/paskino/libace-conda/blob/master/build_ace.sh which uses CMake to download, figure out if it's APPLE or Linux and copies files, and then calls `make`.  It's in the last step that something fails. So, somewhere you should have the source extracted, and can just do `make` there to see what's actually going on. (I guess this should work on OSX as @bathomas did this originally.)

# Confusing about `ACE_DIR`
As far as I can see, `ACE_DIR` isn't used by anyone. CMake [find_package](https://cmake.org/cmake/help/v3.13/command/find_package.html) searches using "Module"-mode (using `FindACE.cmake`) and then falls back on "Config"-mode. The latter uses `ACE_DIR` but will always fail as ACE isn't built with CMake and certainly doesn't write CMake config files. So, setting `ACE_DIR` is never going to help.
(Arguably, Gadgetron should disable the config-mode search for using `find_package(ACE MODULE)` which would avoid this particular confusion).

# `find_package(ACE)` module mode fails

First, are we sure it does? (i.e. does Gadgetron complain and refuse to build?) I would find this weird. We've been building Gadgetron with `brew ACE` for ages now. Indeed, Travis still does. If indeed it doesn't find it, read on...

As there's no `FindACE.cmake` supplied by CMake, Gadgetron comes with its own (always a sign of trouble). As @paskino says, it uses the environment variable `ACE_ROOT`, it does *not* use the CMake variable `ACE_ROOT` (and as above, not `ACE_DIR` either). His advice above is therefore correct.

You can see this [here](https://github.com/gadgetron/gadgetron/blob/b74270142683cb461292e7714f1755969a5bbccf/cmake/FindACE.cmake#L31): it uses [`find_path`](https://cmake.org/cmake/help/v3.13/command/find_path.html) and [`find_library`](https://cmake.org/cmake/help/v3.13/command/find_library.html). Those go and hunt for files in certain standard system locations, and a few ones specified by hand in `FindACE.cmake`.

Maybe this helps you sort it out.

**Comment by KrisThielemans** [comment]
Regarding the build error
```
c++ -fvisibility=hidden -fvisibility-inlines-hidden -Wnon-virtual-dtor -O3 -ggdb -pthread -fno-strict-aliasing -Wall -W -Wpointer-arith -pipe -DACE_LACKS_LINUX_NPTL Unknown -D_GNU_SOURCE   -I/Users/rich/Documents/Code/SIRF-Superbuild/Build/builds/ACE/build/source/ -D__ACE_INLINE__ -I.. -DACE_BUILD_DLL  -c -fPIC -o .shobj/ACE.o /Users/rich/Documents/Code/SIRF-SuperBuild/Build/builds/ACE/build/source/ace/ACE.cpp
clang: error: no such file or directory: 'Unknown'
```
A bit of detective work shows that ACE sets `CPPFLAGS` in its `include/makeinclude` files. I believe the problematic code is [here](https://github.com/DOCGroup/ACE_TAO/blob/95c147c08eaad4f651347ccf898c88b3a5fca705/ACE/include/makeinclude/platform_linux_common.GNU#L110):
```make
PLATFORM_LARGEFILE_CFLAGS := $(shell getconf LFS_CFLAGS 2> /dev/null || echo Unknown)
ifdef PLATFORM_LARGEFILE_CFLAGS
  # Large file support
  CPPFLAGS += $(PLATFORM_LARGEFILE_CFLAGS)
endif #largefile
```
My prediction is @rijobro that if you type
```sh
getconf LFS_CFLAGS
echo $?
```
you'll see some error and a non-zero exit-code. Therefore it sets the variable to `Unknown`. However, that is clearly a bug. It shouldn't set the variable at all (and certainly not to something that the C++ compiler cannot handle).

So, this is an ACE bug as far as I can see. @rijobro, can you create an [ACE issue](https://github.com/DOCGroup/ACE_TAO/issues/new)? (They have a [file describing bugtrackers etc](http://www.dre.vanderbilt.edu/~schmidt/DOC_ROOT/ACE/docs/ACE-bug-process.html) but as it's now on github, I guess an issue is fine).

There's of course the other question why `getconf` failed.

**Comment by rijobro** [comment]
Is this definitely an ACE problem? In filling out their issue form on Github, it asks for the `ace/config.h` file. This symlinks to the platform-specific config file, and mine points to `config-linux.h`, which seems incorrect. 

Has the `USE_SYSTEM_ACE=OFF` worked in the past for OSX? Because looking at Travis, none of the OSX machines test that.

**Comment by KrisThielemans** [comment]
The quoted lines are definitely incorrect.

That it throws up an error now, is not necessarily an ACE problem. You can check what the ACE-install script is doing. In particular, it links the config files in https://github.com/paskino/libace-conda/blob/master/configure_ace.sh, after platform detection by CMake. You could just do this by hand and see what happens. (I tried to track it for OSX, and got to those lines, but possibly I tracked wrongly).

`USE_SYSTEM_ACE=OFF` hasn't been properly tested anywhere. Experimental code put together by @bathomas and @paskino (e.g. for CentOS). I don't know what they tested it on.

**Comment by paskino** [comment]
Other than Gadgetron, I've run no tests. 
Maybe I just have to invert [these](https://github.com/paskino/libace-conda/blob/master/CMakeLists.txt#L27)

**Comment by paskino** [comment]
I updated the [libace-conda CMakeLists.txt](https://github.com/paskino/libace-conda/blob/master/CMakeLists.txt#L27)
Could you update and retry?

**Comment by rijobro** [comment]
That seems to have worked!

**Comment by KrisThielemans** [comment]
I filed https://github.com/DOCGroup/ACE_TAO/issues/792. I guess this can be closed then

**Comment by rijobro** [comment]
Thanks

**Comment by jwillemsen** [comment]
Issue in ACE has been fixed, will be part of ACE 6.5.4
